chia/daemon/keychain_proxy.py

```diff
@@ -89,13 +97,80 @@ def format_request(self, command: str, data: Dict[str, Any]) -> WsRpcMessage:
 
         return super().format_request(command, data)
 
+    async def _get(self, request: WsRpcMessage) -> WsRpcMessage:
+        """
+        Overrides DaemonProxy._get() to handle the connection state
+        """
+        try:
+            if not self.disconnect:  # if we are disconnected, and we send a request we should throw original error.
+                await asyncio.wait_for(self.connection_established.wait(), timeout=10)  # is 10 seconds too long?
+            else:
+                self.log.error("Attempting to send request to a keychain-proxy that has shut down.")
+            self.log.debug(f"Sending request to keychain command: {request['command']} from {request['origin']}.")
+            return await super()._get(request)
+        except asyncio.TimeoutError:
+            raise KeychainProxyConnectionFailure("Could not reconnect to Keychain!")
+
+    async def start(self) -> None:
+        self.keychain_connection_task = asyncio.create_task(self.connect_to_keychain())
+        await self.connection_established.wait()  # wait until connection is established.
+
+    async def connect_to_keychain(self) -> None:
+        while not self.disconnect:
+            try:
+                self.client_session = ClientSession()
+                self.websocket = await self.client_session.ws_connect(
+                    self._uri,
+                    autoclose=True,
+                    autoping=True,
+                    heartbeat=60,
+                    ssl_context=self.ssl_context,
+                    max_msg_size=self.max_message_size,
+                )
+                await self.connection_listener()
+            except ClientConnectorError:
+                self.log.warning(f"Can not connect to keychain at {self._uri}.")
+            except Exception as e:
+                tb = traceback.format_exc()
+                self.log.warning(f"Exception: {tb} {type(e)}")
+            self.connection_established.clear()
+            if self.websocket is not None:
+                await self.websocket.close()
+            if self.client_session is not None:
+                await self.client_session.close()
+            self.websocket = None
+            self.client_session = None
+            await asyncio.sleep(2)
+
+    async def connection_listener(self) -> None:
```

Comment:
The logic in here seems basically identical to DaemonProxy's listener() method, with the exception of calling `self.connection_established.set()` and the "Close signal received..." log message. Can we leverage super.listener() to avoid duplicating code?

Reply:
it calls self.close so we sadly can't

Reply:
Maybe we can move the common part into a new method? This part specifically:

```                if message.type == WSMsgType.TEXT:
                    decoded: WsRpcMessage = json.loads(message.data)
                    request_id = decoded["request_id"]

                    if request_id in self._request_dict:
                        self.response_dict[request_id] = decoded
                        self._request_dict[request_id].set()```

Reply:
done